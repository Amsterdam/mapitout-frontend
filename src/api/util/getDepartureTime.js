export const getDepartureTime = (date = new Date()) => {
  const dayOfWeek = date.getUTCDay();

  if (dayOfWeek !== 1) {
    const dayOfMonth = date.getUTCDate();

    date.setUTCDate(dayOfMonth - dayOfWeek + 8);
  }

  date.setUTCHours(9, 0, 0, 0);

  return date;
};
