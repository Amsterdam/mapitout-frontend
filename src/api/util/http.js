export const request = async (url, requestConfig) => {
  let response;

  response = await fetch(url, requestConfig);

  if (response.ok) {
    try {
      return await response.json();
    } catch (error) {
      throw new Error("Invalid server response");
    }
  } else {
    throw new Error("Invalid server response");
  }
};
