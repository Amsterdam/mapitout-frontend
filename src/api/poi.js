import { request } from "./util/http";
import { ENDPOINT_POI_SEARCH } from "../env";
import { isArray } from "lodash-es";

const requestConfig = {
  method: "POST",
  headers: {
    Accept: "application/json",
    "Content-type": "application/json; charset=utf-8"
  }
};

export const fetchPois = async (
  area = { paths: [] },
  types = [],
  subTypes = []
) => {
  if (
    typeof area !== "object" ||
    !area.hasOwnProperty("paths") ||
    !isArray(area.paths) ||
    area.paths.length === 0
  ) {
    return [];
  }
  const requestBody = {
    poi_in_polygon: {
      type: "Feature",
      geometry: {
        type: "MultiPolygon",
        coordinates: area.paths.map(polygon => [
          polygon.map(point => [point.lng, point.lat])
        ]),
        crs: {
          type: "name",
          properties: {
            name: "EPSG:4326"
          }
        }
      }
    }
  };

  if (types.length > 0) {
    requestBody.poi_by_type = types;
  }

  if (subTypes.length > 0) {
    requestBody.poi_by_property = subTypes;
  }

  const url = new URL(ENDPOINT_POI_SEARCH);

  const response = await request(url.toString(), {
    ...requestConfig,
    body: JSON.stringify(requestBody)
  });

  return response.map(locationData => locationData[0]);
};

export const lookupPoi = async (poiName = "") => {
  if (typeof poiName !== "string" || poiName.length === 0) {
    return null;
  }
  const url = new URL(ENDPOINT_POI_SEARCH);

  const response = await request(url.toString(), {
    ...requestConfig,
    body: JSON.stringify({
      poi_by_name: poiName
    })
  });

  if (response[0][0]) {
    return {
      name: response[0][0].name,
      description: response[0][0].description,
      address: `${response[0][0].street}, ${response[0][0].postalcode} ${
        response[0][0].city
      }`,
      website: response[0][0].website,
      phone: response[0][0].phone,
      lng: response[0][0].geo_location.coordinates[0],
      lat: response[0][0].geo_location.coordinates[1],
      typeId: response[0][0].poi_type_id
    };
  }

  return null;
};
