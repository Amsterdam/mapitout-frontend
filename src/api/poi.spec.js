import { fetchPois, lookupPoi } from "./poi";
import * as http from "./util/http";

afterEach(() => {
  jest.resetAllMocks();
});

jest.mock("../env.js");

const requestSpy = jest.spyOn(http, "request").mockImplementation();

describe("fetchPois", () => {
  it("should return an empty array when passed an invalid or empty area parameter", async () => {
    const result1 = await fetchPois({});
    const result2 = await fetchPois({
      paths: {}
    });
    const result3 = await fetchPois({
      paths: []
    });
    const result4 = await fetchPois();

    expect(result1).toEqual([]);
    expect(result2).toEqual([]);
    expect(result3).toEqual([]);
    expect(result4).toEqual([]);
    expect(requestSpy).toHaveBeenCalledTimes(0);
  });

  it("should call request with the correct url and request configuration object", async () => {
    const testArea = {
      paths: [[{ lat: 1, lng: 2 }], [{ lat: 3, lng: 4 }]]
    };
    const testTypes = [1];
    const testSubTypes = [2];

    requestSpy.mockResolvedValue([]);

    await fetchPois(testArea);
    await fetchPois(testArea, testTypes);
    await fetchPois(testArea, [], testSubTypes);
    await fetchPois(testArea, testTypes, testSubTypes);

    expect(requestSpy).toHaveBeenCalledTimes(4);

    expect(requestSpy.mock.calls[0]).toMatchSnapshot();
    expect(requestSpy.mock.calls[1]).toMatchSnapshot();
    expect(requestSpy.mock.calls[2]).toMatchSnapshot();
    expect(requestSpy.mock.calls[3]).toMatchSnapshot();
  });

  it("should throw an error when request rejects", async () => {
    const testArea = {
      paths: [[{ lat: 1, lng: 2 }], [{ lat: 3, lng: 4 }]]
    };

    const expectedError = new Error("test-error");

    requestSpy.mockRejectedValue(expectedError);

    await expect(fetchPois(testArea)).rejects.toThrow(expectedError);
  });

  it("should throw an error if request returns an invalid response", async () => {
    const testArea = {
      paths: [[{ lat: 1, lng: 2 }], [{ lat: 3, lng: 4 }]]
    };

    requestSpy.mockResolvedValue({});

    await expect(fetchPois(testArea)).rejects.toThrow();
  });

  it("should return an empty array if request returns an empty response", async () => {
    const testArea = {
      paths: [[{ lat: 1, lng: 2 }], [{ lat: 3, lng: 4 }]]
    };

    const expectedResult = [];

    requestSpy.mockResolvedValue([]);

    const result = await fetchPois(testArea);

    expect(result).toEqual(expectedResult);
  });

  it("should return the correct value when request resolves with a valid response", async () => {
    const testArea = {
      paths: [[{ lat: 1, lng: 2 }], [{ lat: 3, lng: 4 }]]
    };

    const expectedResult = [{ test: "test" }];

    requestSpy.mockResolvedValue([[{ test: "test" }]]);

    const result = await fetchPois(testArea);

    expect(result).toEqual(expectedResult);
  });
});

describe("lookupPoi", () => {
  const validResponse = {
    name: "test",
    description: "test",
    street: "test street",
    postalcode: "1111 GH",
    city: "test city",
    website: "http://www.example.com",
    phone: "123456",
    geo_location: {
      coordinates: [1, 2]
    },
    typeId: 0
  };
  it("should return null when passed an invalid or empty name parameter", async () => {
    const result1 = await lookupPoi({});
    const result2 = await lookupPoi("");

    expect(result1).toEqual(null);
    expect(result2).toEqual(null);
    expect(requestSpy).toHaveBeenCalledTimes(0);
  });

  it("should call request with the correct url and request configuration object", async () => {
    const testPoiName = "test";

    requestSpy.mockResolvedValue([[validResponse]]);

    await lookupPoi(testPoiName);

    expect(requestSpy).toHaveBeenCalledTimes(1);
    expect(requestSpy.mock.calls[0]).toMatchSnapshot();
  });

  it("should throw an error when request rejects", async () => {
    const testPoiName = "test";

    const expectedError = new Error("test-error");

    requestSpy.mockRejectedValue(expectedError);

    await expect(lookupPoi(testPoiName)).rejects.toThrow(expectedError);
  });

  it("should throw an error when request returns an invalid response", async () => {
    const testPoiName = "test";

    requestSpy.mockResolvedValue({});

    await expect(lookupPoi(testPoiName)).rejects.toThrow();
  });

  it("should return null if request returns an empty response", async () => {
    const testPoiName = "test";

    const expectedResult = null;

    requestSpy.mockResolvedValue([[]]);

    const result = await lookupPoi(testPoiName);

    expect(result).toEqual(expectedResult);
  });

  it("should return the correct value when request resolves with a valid response", async () => {
    const testPoiName = "test";
    const expectedResult = {
      address: "test street, 1111 GH test city",
      description: "test",
      lat: 2,
      lng: 1,
      name: "test",
      phone: "123456",
      typeId: undefined,
      website: "http://www.example.com"
    };

    requestSpy.mockResolvedValue([
      [
        {
          name: "test",
          description: "test",
          street: "test street",
          postalcode: "1111 GH",
          city: "test city",
          website: "http://www.example.com",
          phone: "123456",
          geo_location: {
            coordinates: [1, 2]
          },
          typeId: 0
        }
      ]
    ]);

    const result = await lookupPoi(testPoiName);

    expect(result).toEqual(expectedResult);
  });
});
