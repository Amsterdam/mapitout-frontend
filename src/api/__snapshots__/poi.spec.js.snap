// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchPois should call request with the correct url and request configuration object 1`] = `
Array [
  "https://www.example.com/poi-search",
  Object {
    "body": "{\\"poi_in_polygon\\":{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"MultiPolygon\\",\\"coordinates\\":[[[[2,1]]],[[[4,3]]]],\\"crs\\":{\\"type\\":\\"name\\",\\"properties\\":{\\"name\\":\\"EPSG:4326\\"}}}}}",
    "headers": Object {
      "Accept": "application/json",
      "Content-type": "application/json; charset=utf-8",
    },
    "method": "POST",
  },
]
`;

exports[`fetchPois should call request with the correct url and request configuration object 2`] = `
Array [
  "https://www.example.com/poi-search",
  Object {
    "body": "{\\"poi_in_polygon\\":{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"MultiPolygon\\",\\"coordinates\\":[[[[2,1]]],[[[4,3]]]],\\"crs\\":{\\"type\\":\\"name\\",\\"properties\\":{\\"name\\":\\"EPSG:4326\\"}}}},\\"poi_by_type\\":[1]}",
    "headers": Object {
      "Accept": "application/json",
      "Content-type": "application/json; charset=utf-8",
    },
    "method": "POST",
  },
]
`;

exports[`fetchPois should call request with the correct url and request configuration object 3`] = `
Array [
  "https://www.example.com/poi-search",
  Object {
    "body": "{\\"poi_in_polygon\\":{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"MultiPolygon\\",\\"coordinates\\":[[[[2,1]]],[[[4,3]]]],\\"crs\\":{\\"type\\":\\"name\\",\\"properties\\":{\\"name\\":\\"EPSG:4326\\"}}}},\\"poi_by_property\\":[2]}",
    "headers": Object {
      "Accept": "application/json",
      "Content-type": "application/json; charset=utf-8",
    },
    "method": "POST",
  },
]
`;

exports[`fetchPois should call request with the correct url and request configuration object 4`] = `
Array [
  "https://www.example.com/poi-search",
  Object {
    "body": "{\\"poi_in_polygon\\":{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"MultiPolygon\\",\\"coordinates\\":[[[[2,1]]],[[[4,3]]]],\\"crs\\":{\\"type\\":\\"name\\",\\"properties\\":{\\"name\\":\\"EPSG:4326\\"}}}},\\"poi_by_type\\":[1],\\"poi_by_property\\":[2]}",
    "headers": Object {
      "Accept": "application/json",
      "Content-type": "application/json; charset=utf-8",
    },
    "method": "POST",
  },
]
`;

exports[`lookupPoi should call request with the correct url and request configuration object 1`] = `
Array [
  "https://www.example.com/poi-search",
  Object {
    "body": "{\\"poi_by_name\\":\\"test\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Content-type": "application/json; charset=utf-8",
    },
    "method": "POST",
  },
]
`;
